### PR TITLE
Improve startup / relaunch screens

### LINKED ISSUE #384: Google app automatic backup restore, give option to re-download modules that user had before
Related to #365 in which the Google app automatic backup is fixed.

When reinstalling AB after a uninstall, or installing on a new phone, Google will (after #365 is closed) restore the app settings, notes, and bookmarks. This is if Google's "Back up to Google Drive" feature is enabled which is available from Android 6.0

**Feature request**
Save downloaded module names in shared preferences, so that on a new install AB will ask user if he/she wants to download the missing modules, or even be able to unselect some if they don't want to install all the modules they used before.

### LINKED ISSUE #797: Improve cold start experience
When And Bible is closed (removed from task list), it can easily take from 3 to 6 seconds to load.
On my one phone (night mode), it takes about 3 seconds and goes like this:
1. .5 second full white screen
2. 1.5 seconds full black screen
3. Quick flash of startup screen
4. Bible screen comes up with brief black background while text loads
5. Fully loaded

In non-night mode (white mode), steps 2 and 3 are not happening but step 1 seems to take that much longer
Also, with #663 theme improvement, steps 2 and 3 are not happening for either night or non-night mode, but start up time is still the same, with step 1 taking longer.

**My suggestion:**
Build a splash screen with loading indicator to show on cold start. When app is in background and re-opening or switching back to it, splash screen should never show.

### LINKED ISSUE #877: Remove blue startup screen
I believe startup screen was made because initializing jsword took time in the past.
I also believe that it is these days neglible in most cases. User suggested removing it or making int optional and feels it's disturbing.

### PR DESCRIPTION
**Pull request by RuggMatt**
Improves first start experience, and relaunch experience

On loading the app for the first time, user sees a screen like:

__Dark theme__

__Light theme__

If launching the app when modules are installed, the app shows a spinner while loading. 

Some work towards #798. 
- [x] Button to download bibles
- [x] Button to import modules
- [x] Redownload previous Bibles, if any. 
  - [x] Created DB schema to store list of modules
  - [x] DB is populated with currently installed modules when user updates app
  - [x] Removing a module removes it form the DB
  - [x] Installing a new module adds it to the DB
  - [x] User sees a button on a new install if DB is restored (from Google Backup or DB restore)
  - [x] Clicking that button gives the user a checklist of which modules to attempt to install 
  - [x] Checked list of modules is downloaded
  - [x] Any modules which fail to download are displayed in a list to the user. 

Closes #384
Closes #797
Closes #877 

Related  #294

### CODE REVIEW

**Comment by tuomas2** [comment]
Looks good so far. Though I did not yet look at the code. 

>  Redownload previous Bibles, if any.

Do you have a plan of doing this? How do you figure out the documents that were installed? By allowing user to restore
database backup at this point?

**Comment by RuggMatt** [comment]
I'm thinking something like this. If user is reinstalling the app, Android will restore it's DB too if user has chosen to backup app data. This currently includes bookmarks and notes, among other things.
- When using the app, add code to generate and save info about currently installed books into the DB (osisId, where it was downloaded from, etc). The info about each document will be backed up, and will also need to be updated whenever a user downloads, installs, or removes a document.
- On first launch/reinstall, check DB to see if it says any documents were previously installed. Check if any actually are. If there are none, but the DB says there is, prompt the user to re-download documents that are still available (not all may be, (e.g. ESV)).

**Comment by timbze** [comment]
Looks good! Though I haven't looked at all the code in much detail yet either.

It makes sense that sword warm up in application start is no longer needed, cause phones have become so much faster. But can you just briefly give what you found or why you removed it? Just want to be sure nothing will break, though I don't think it should.

A small issue, the loading screen with spinner has theme based on device and not And Bible setting. My phone is on night mode but when I manually set And Bible to day mode the start screen with spinner is still dark. It should base on AB setting and not just system.

**Comment by RuggMatt** [comment]
> But can you just briefly give what you found or why you removed it? 

The only thing the warm up code did was set a timer task for 3 seconds and then make a call to `SwordDocumentFacade#getBibles()`, which was already happening anyway in the `StartupActivity`, as it checks to see whether to continue to the main view, or to prompt to download Bibles. 

Added some more things:
- If a user reinstalls the app/installs on a new device, but had previously installed this version or later and had it backed up to Google, then they will see a third button to redownload documents.


- Choosing this option, they will see a list of books they previously had installed, and a way to select which ones they want to attempt to download



- After selecting them, they are redirected to the download activity, which looks for and queues the documents for download, if they are found:


Other ideas not yet implemented:
- [x] Show a notice to the user if any modules couldn't be found that they previously had installed (right now it just silently ignores if doc not found)
- [x] Still trying to figure out where this light/dark theme inconsistency is coming from.

**Comment by timbze** [comment]
@mattschlosser thanks for the info on warm up code. Sounds good.

From functionality perspective I think you're on point. Very good. 

From UX perspective I think there could be some improvement. "Quick Redownload" does not really give me an idea what it's about unless I know it somehow.

I think perhaps we could have a message under the AB logo that would give a short welcome / next steps note. For example:
***
Thank you for downloading Bible Study app by And Bible Open Source Project.

There are currently no Bibles or documents installed. To continue, please install at least 1 document by downloading from the internet, or by loading from a zip file document backup. {The database (restored from Google backup) has a record of documents you have had installed previously. If you want to install those documents please click the "Quick Redownload" button below.}
***

Message in brackets {} would only show when "Quick Redownload" button shows. I have not thought of message for very long, perhaps there are better ways of saying it, but I think that would make it much friendlier for many users.

**Comment by tuomas2** [comment]
I'm also thinking that in addition to automatic google backup (does it work?), we should support restoring manual backup at this point, and based on that, suggest user to download missing documents. Though, he might want to use his own document backup zip as well.

**Comment by timbze** [comment]
> I'm also thinking that in addition to automatic google backup (does it work?), we should support restoring manual backup at this point, and based on that, suggest user to download missing documents. Though, he might want to use his own document backup zip as well.

I agree. Yes I think Auto Google backup works. Have not tested lately, but not too long ago I have.

**Comment by RuggMatt** [comment]
> I'm also thinking that in addition to automatic google backup (does it work?), we should support restoring manual backup at this point, and based on that, suggest user to download missing documents. Though, he might want to use his own document backup zip as well.

Good idea. Perhaps the third button could vary depending on if Google restored the DB or not. 

> > I'm also thinking that in addition to automatic google backup (does it work?), we should support restoring manual backup at this point, and based on that, suggest user to download missing documents. Though, he might want to use his own document backup zip as well.
> 
> I agree. Yes I think Auto Google backup works. Have not tested lately, but not too long ago I have.

Yes, backup works to my knowledge. I have been testing this, and it works great. 

> @mattschlosser thanks for the info on warm up code. Sounds good.
> 
> From functionality perspective I think you're on point. Very good.
> 
> From UX perspective I think there could be some improvement. "Quick Redownload" does not really give me an idea what it's about unless I know it somehow.
> 
> I think perhaps we could have a message under the AB logo that would give a short welcome / next steps note. For example:
> 
> Thank you for downloading Bible Study app by And Bible Open Source Project.
> 
> There are currently no Bibles or documents installed. To continue, please install at least 1 document by downloading from the internet, or by loading from a zip file document backup. {The database (restored from Google backup) has a record of documents you have had installed previously. If you want to install those documents please click the "Quick Redownload" button below.}
> 
> Message in brackets {} would only show when "Quick Redownload" button shows. I have not thought of message for very long, perhaps there are better ways of saying it, but I think that would make it much friendlier for many users.

Thanks for the suggestions.

**Comment by timbze** [comment]
Would you make the links at the bottom clickable yet? Not sure if necessary but I think that would be nice

**Comment by timbze** [comment]
Since this will be merged after vue branch, I'm switching to merge to vue branch so that conflicts can be fixed. If @mattschlosser can fix that's great, if not I'll probably do it within a few days when I install vue branch on my phone.

**Comment by RuggMatt** [comment]
@timbze I was planning to rebase from develop after the merge, but this works too.

**Comment by timbze** [comment]
I've merged vue branch here now. If you already have this branch installed on your phone you'll need to make the following change in the code for the first startup:
In DatabaseContainer.kt, migration 35-36 add the code from migration 34-35 which is:
```kotlin
db.execSQL("ALTER TABLE `Bookmark` ADD COLUMN `book` TEXT")
db.execSQL("ALTER TABLE `Bookmark` ADD COLUMN `startOffset` INTEGER DEFAULT NULL")
db.execSQL("ALTER TABLE `Bookmark` ADD COLUMN `endOffset` INTEGER DEFAULT NULL")
```

**Comment by timbze** [comment]
@mattschlosser do you have migration code so that existing installed documents will be saved to database when user upgrades to this app version?

**Comment by RuggMatt** [comment]
> 
> 
> @mattschlosser do you have migration code so that existing installed documents will be saved to database when user upgrades to this app version?

Yes. There is a [function in MainBibleActivity.kt](https://github.com/AndBible/and-bible/blob/6354139222c90f096e9274332bb75cf9deed13c0/app/src/main/java/net/bible/android/view/activity/page/MainBibleActivity.kt#L311) that does this. It checks if the Downloaded Documents DB is empty and loads it with the docs  currently installed.

**Comment by timbze** [comment]
> Yes. There is a [function in MainBibleActivity.kt](https://github.com/AndBible/and-bible/blob/6354139222c90f096e9274332bb75cf9deed13c0/app/src/main/java/net/bible/android/view/activity/page/MainBibleActivity.kt#L311) that does this. It checks if the Downloaded Documents DB is empty and loads it with the docs currently installed.

It seems like this code will only be useful once. In that case if might be better to use upgradeSharedPreferences() in BibleApplication.kt. Or am I missing something?

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -651,6 +653,17 @@
         other extra material when you study bible.
     </string>
 
+    <!-- Welcome text -->
+    <string name="redownload_documents">Quick Redownload</string>
+    <string name="welcome_message">Thank you for downloading Bible Study app by And Bible Open Source
+        Project. There are currently no Bibles or documents installed. To continue, please install
+        at least 1 document by downloading from the internet, or by loading from a zip file document
```
Comment (tuomas2) [inline app/src/main/res/values/strings.xml:660]:
```suggestion
        at least 1 Bible document by downloading from the internet, or by loading from a zip file document
```

File: app/src/main/java/net/bible/service/db/DatabaseContainer.kt [hunk 2]
```diff
@@ -744,6 +744,14 @@ private val JOURNAL_39_40 = object : Migration(39, 40) {
     }
 }
 
+private val MIGRATION_40_41_DocumentBackup = object : Migration(40, 41) {
+    override fun doMigrate(db: SupportSQLiteDatabase) {
+        db.apply {
+            execSQL("""CREATE TABLE IF NOT EXISTS `DocumentBackup` (`osisId` TEXT PRIMARY KEY NOT NULL, `abbreviation` TEXT NOT NULL, `name` TEXT NOT NULL, `language` TEXT NOT NULL, `repository` TEXT NOT NULL);""")
```
Comment (tuomas2) [inline app/src/main/java/net/bible/service/db/DatabaseContainer.kt:750]: Why not putting currently installed docs to db here?

Comment (tuomas2) [reply app/src/main/java/net/bible/service/db/DatabaseContainer.kt:750]: i.e. what is done in checkDocBackupDBInSync

File: app/src/main/java/net/bible/android/view/activity/page/MainBibleActivity.kt [hunk 3]
```diff
@@ -300,6 +303,32 @@ class MainBibleActivity : CustomTitlebarActivityBase() {
             }
         }
         initialized = true
+        GlobalScope.launch {
+            checkDocBackupDBInSync()
+        }
```
Comment (tuomas2) [inline app/src/main/java/net/bible/android/view/activity/page/MainBibleActivity.kt:308]: I think this could be done in the above if(!initialized) block

**Review by tuomas2 — APPROVED** [review]
Very good work and beautifully done! Thank you! I left comments, you can create further PRs if you are addressing them. Merging now.

BTW sorry that I didn't review this earlier. I didn't notice that it was changed from draft to ready for review.

**Pull request merged**